### ISSUE TITLE
samples cannot be dumped but must be manually written.

### ISSUE THREAD

**Issue by yvan** [SMAPPNYU/pysmap#22]
they should be dumpable at least for dataests.

**Comment by yvan** [comment]
fixed by replacing the get iterator method of a copy of the collection or dataset with a function that yields tweets via reservoir sampling.

edit:

it should be fixed for a collection (on the branch, not in production) but the data set class needs many methods to be rewritten.

**Comment by yinleon** [comment]
@yvan  I think this is the same problem that occured for Dennis and Megan.

**Comment by yvan** [comment]
it is the same issue. i was working on fixing it, temporary workaround:

```python
with open('file_dump.json', 'w') as f:
        for tweet in collection.sample(10):
	    f.write(json.dumps(tweet))
	    f.write('\n')
```

**Comment by yvan** [comment]
```
 # the reason this implementation doesnt work is because 
 # it doestn access the get_collection_iterators
 # which is what get overidden by the reservoir sample.
        for i, collection in enumerate(self.collections):
            if parallel:
                filename, file_extension = output_file.split(os.extsep, 1)
                collection.dump_to_json('{}_{}.{}'.format(filename, i, file_extension))
            else:
                collection.dump_to_json(output_file)
```

```
# this implemetnation fails to work for the same reason

        if parallel:
            filename, file_extension = output_file.split(os.extsep, 1)
            for i,collection in enumerate(self.collections):
                filehandle = open('{}_{}.{}'.format(filename, i, file_extension), 'ab+')
                for tweet in collection.get_iterator():
                    filehandle.write(BSON.encode(tweet))
                filehandle.close()
        else:
            filehandle = open(output_file, 'ab+')
            for collection in self.collections:
                for tweet in collection.get_iterator():
                    filehandle.write(BSON.encode(tweet))
            filehandle.close()
```

**Comment by yvan** [comment]
we cannot keep support for parallel dump option in its current form. when we override the get_collection_iterators with the function taht yield individual elements of a sample, we lose access to where those samples came from. i will try to make parallel wirte a fixed # of files (we will leave it to the user to decide how many files tow write out).

**Comment by yvan** [comment]
so far tried:

```
        # for i, collection in enumerate(self.collections):
        #     if parallel:
        #         filename, file_extension = output_file.split(os.extsep, 1)
        #         collection.dump_to_bson('{}_{}.{}'.format(filename, i, file_extension))
        #     else:
        #         collection.dump_to_bson(output_file)

        # if parallel:
        #     for i in range(0, parallel):
        #         filename, file_extension = output_file.split(os.extsep, 1)
        #         filehandle = open(output_file, 'ab+')
        #         for tweet in self.collection.get_collection_iterators():
        #             filehandle.write(BSON.encode(tweet))
        #         filehandle.close()
        #         # collection.dump_to_bson('{}_{}.{}'.format(filename, i, file_extension))
        # else:
        #     collection.dump_to_bson(output_file)

        # if parallel:
        #     filename, file_extension = output_file.split(os.extsep, 1)
        #     for i,collection in enumerate(self.collections):
        #         filehandle = open('{}_{}.{}'.format(filename, i, file_extension), 'ab+')
        #         for tweet in collection.get_iterator():
        #             filehandle.write(BSON.encode(tweet))
        #         filehandle.close()
        # else:
        #     filehandle = open(output_file, 'ab+')
        #     for collection in self.collections:
        #         for tweet in collection.get_iterator():
        #             filehandle.write(BSON.encode(tweet))
        #     filehandle.close()
```

**Comment by yvan** [comment]
it looks like we are repeating a lot of code that should not be repeated but the subtlety is in the iterator call when we loop over it. we cant just offload the implementation somewhere else because passing around iterators is odd. may choose to reconsolidate in future, but first i want to make sure the structure works and is usable across single collections and multi collection datasets first.

**Comment by yvan** [comment]
the base issue is fixed in 0.0.37 although it still wont work with filters AFTER sample is called, i am adding that support soon. so the only thing you can call after sample is a dump method.

so queries should be written in the form: 
dataset.filter1().filter2().sample(n).dump_to_json('out.json')

i also fixed a bug in sqlite dump that would chomp off the last few thousand tweets under some circumstances.
